modules/core/src/test/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataRemoveTest.java

```diff
@@ -0,0 +1,308 @@
+    /** Max retry cont. */
+    private static final int MAX_RETRY_CONT = 10;
+
+    /** */
+    private static final String CACHE_NAME = "cache";
+
+    /** */
+    private GridTestUtils.DiscoveryHook discoveryHook;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        TcpDiscoverySpi discoSpi;
+
+        final GridTestUtils.DiscoveryHook discoveryHook0 = discoveryHook;
+
+        if (discoveryHook0 != null) {
+            discoSpi = new TcpDiscoverySpi() {
+                @Override public void setListener(@Nullable DiscoverySpiListener lsnr) {
+                    if (discoveryHook0 != null)
+                        super.setListener(GridTestUtils.DiscoverySpiListenerWrapper.wrap(lsnr, discoveryHook0));
+                }
+            };
+        }
+        else
+            discoSpi = new TcpDiscoverySpi();
+
+        cfg.setDiscoverySpi(discoSpi);
+
+        cfg.setCacheConfiguration(new CacheConfiguration().setName(CACHE_NAME));
+
+        return cfg;
+    }
+
+    /**
+     *
+     */
+    protected void startCluster() throws Exception {
+        startGrid("srv0");
+        startGrid("srv1");
+        startGrid("srv2");
+        startClientGrid("cli0");
+        startClientGrid("cli1");
+        startClientGrid("cli2");
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        startCluster();
+
+        discoveryHook = null;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+    }
+
+    /**
+     * Tests remove not existent type and checks the exception.
+     */
+    @Test
+    public void testRemoveNotExistentType() {
+        for (Ignite testNode : G.allGrids()) {
+            GridTestUtils.assertThrows(log, () -> {
+                    ((IgniteEx)testNode).context().cacheObjects().removeType(
+                        ((IgniteEx)testNode).context().cacheObjects().typeId("NotExistentType"));
+
+                    return null;
+                },
+                IgniteException.class, "Failed to remove metadata, type not found");
+        }
+    }
+
+    /**
+     * Tests remove type metadata at all nodes (coordinator, server, client).
+     */
+    @Test
+    public void testRemoveTypeOnNodes() throws Exception {
+        List<IgniteEx[]> testNodeSets = new ArrayList<>();
+
+        // Add all servers permutations to tests sets.
+        for (Ignite ign0 : G.allGrids()) {
+            for (Ignite ign1 : G.allGrids()) {
+                for (Ignite ign2 : G.allGrids()) {
+                    IgniteEx ignx0 = (IgniteEx)ign0;
+                    IgniteEx ignx1 = (IgniteEx)ign1;
+                    IgniteEx ignx2 = (IgniteEx)ign2;
+
+                    if (!ignx0.context().clientNode()
```

Comment:
Test description says that removal should be verified on client nodes too, but this condition seems exclude client nodes at all

Reply:
Sure, take a look below.
We check all combinations of server nodes and add test set with  client nodes manually.